### PR TITLE
Broker leadership evacuation

### PR DESCRIPTION
**Pull request by zrosenberg**
Adds the functionality to the `rebuild` command only, to evacuate leadership for given broker/topic pairings (easy future functionality for partition granularity). We currently have a few use cases where this is desirable:

- A single broker is overloaded because of partition imbalance by something not currently handled by topicmappr (requests / sec for example). We want to remove load from this broker, as fast as possible, without causing a reassignment and the network overhead that would incur. Swapping leader assignments allows a replica to take over instantly, without needing to move any data. 
- Something has gone wrong in a single AZ, and we want to remove leadership from all brokers in that AZ. By using registry tags, we can quickly make an api call to get a list of all brokers in the problem AZ, and pass that in here. This allows us to offload traffic from the problem AZ, and again requires no data to be shuffled around.

Expected use case is to use `topicmappr rebuild` with the current set of brokers and topics, and additionally pass in `--leader-evac-brokers="10001,10002" --leader-evac-topics=".*"`

I made the topic specification work the same as the current topic regex, in the case where you want to move traffic for a given topic only off a broker.

This was originally going to be a new cmd, but implementation shifted, so the commit history is not particularly helpful. Sorry.

Open question: Do I need to add some more output logging to make it clear to the user what is happening, as we change leadership around?

### CODE REVIEW

**Comment by zrosenberg** [comment]
And right after posting this, we have a use case for partition level granularity as well - when new consumer workloads are being spun up that put some brokers over the edge because of read amplification, partition level leadership changes can be used to easily rebalance network outbound without triggering a reassignment.

File: cmd/topicmappr/commands/reassignments.go [hunk 1]
```diff
@@ -129,3 +131,57 @@ func computeReassignmentBundles(params computeReassignmentBundlesParams) chan re
...
+	// evacuation algo starts here
+	partitionMapOut := partitionMapIn.Copy()
+
+	// make a lookup map of topics
+	topicsMap := map[string]struct{}{}
+	for _, topic := range evacTopics {
+		topicsMap[topic] = struct{}{}
+	}
+
+	// make a lookup map of topics
```
Comment (jamiealquiza) [inline cmd/topicmappr/commands/reassignments.go:148]: dupe comment as L142

File: cmd/topicmappr/commands/reassignments.go [hunk 2]
```diff
@@ -129,3 +131,57 @@ func computeReassignmentBundles(params computeReassignmentBundlesParams) chan re
...
+
+		// check the leader to see if its one of the evac brokers
+		if _, contains := brokersMap[p.Replicas[0]]; !contains {
+			continue
+		}
+
+		for j, replica := range p.Replicas {
+			// If one of the replicas is not being leadership evacuated, use that one and swap.
+			if _, contains := brokersMap[replica]; !contains {
+				newLeader := p.Replicas[j]
```
Comment (jamiealquiza) [inline cmd/topicmappr/commands/reassignments.go:172]: synonymous with the `replicas` variable in each loop, for whatever it's worth

**Comment by jamiealquiza** [comment]
> And right after posting this, we have a use case for partition level granularity as well - when new consumer workloads are being spun up that put some brokers over the edge because of read amplification, partition level leadership changes can be used to easily rebalance network outbound without triggering a reassignment.

I think if we have many leadership evac conditions to account for and need to add more flags for increased granularity, it might be worth breaking this into a dedicated sub-command

**Comment by zrosenberg** [comment]
@jamiealquiza Happy to make this a separate command, that was my motivation for putting the functionality in its own function. Would you mind if I did that in a separate followup PR though? Don't want to get too blocked on this, since we have a need for it in a gameday.

**Comment by zrosenberg** [comment]
Tried this out, and it works

> ./topicmappr rebuild --brokers '10000, 10001, 10002, 10003, 10004, 10005' --topics loadgen-1 --leader-evac-brokers 10001 --leader-evac-topics .*
> 
> Topics:
>   loadgen-1
> 
> Broker change summary:
>   Replacing 0, added 0, missing 0, total count changed by 0
> 
> Action:
>   no-op
> 
> Partition map changes:
>   loadgen-1 p0: [10005 10003] -> [10005 10003] no-op
>   loadgen-1 p1: [10002 10004] -> [10002 10004] no-op
>   loadgen-1 p2: [10000 10001] -> [10000 10001] no-op
>   loadgen-1 p3: [10001 10003] -> [10003 10001] preferred leader
>   loadgen-1 p4: [10003 10004] -> [10003 10004] no-op
>   loadgen-1 p5: [10004 10000] -> [10004 10000] no-op
>   loadgen-1 p6: [10005 10003] -> [10005 10003] no-op
>   loadgen-1 p7: [10002 10000] -> [10002 10000] no-op
>   loadgen-1 p8: [10000 10005] -> [10000 10005] no-op
>   loadgen-1 p9: [10001 10002] -> [10002 10001] preferred leader
>   loadgen-1 p10: [10003 10005] -> [10003 10005] no-op
>   loadgen-1 p11: [10004 10002] -> [10004 10002] no-op
>   loadgen-1 p12: [10005 10004] -> [10005 10004] no-op
>   loadgen-1 p13: [10002 10001] -> [10002 10001] no-op
>   loadgen-1 p14: [10000 10004] -> [10000 10004] no-op
>   loadgen-1 p15: [10001 10000] -> [10000 10001] preferred leader
>   loadgen-1 p16: [10003 10002] -> [10003 10002] no-op
>   loadgen-1 p17: [10004 10005] -> [10004 10005] no-op
>   loadgen-1 p18: [10005 10001] -> [10005 10001] no-op
>   loadgen-1 p19: [10002 10003] -> [10002 10003] no-op
>   loadgen-1 p20: [10000 10002] -> [10000 10002] no-op
>   loadgen-1 p21: [10001 10003] -> [10003 10001] preferred leader
>   loadgen-1 p22: [10003 10004] -> [10003 10004] no-op
>   loadgen-1 p23: [10004 10000] -> [10004 10000] no-op
>   loadgen-1 p24: [10005 10001] -> [10005 10001] no-op
>   loadgen-1 p25: [10002 10000] -> [10002 10000] no-op
>   loadgen-1 p26: [10000 10005] -> [10000 10005] no-op
>   loadgen-1 p27: [10001 10002] -> [10002 10001] preferred leader
>   loadgen-1 p28: [10003 10005] -> [10003 10005] no-op
>   loadgen-1 p29: [10004 10003] -> [10004 10003] no-op
>   loadgen-1 p30: [10005 10004] -> [10005 10004] no-op
>   loadgen-1 p31: [10002 10001] -> [10002 10001] no-op
>   loadgen-1 p32: [10000 10004] -> [10000 10004] no-op
>   loadgen-1 p33: [10001 10000] -> [10000 10001] preferred leader
>   loadgen-1 p34: [10003 10002] -> [10003 10002] no-op
>   loadgen-1 p35: [10004 10005] -> [10004 10005] no-op
>   loadgen-1 p36: [10005 10001] -> [10005 10001] no-op
>   loadgen-1 p37: [10002 10003] -> [10002 10003] no-op
>   loadgen-1 p38: [10000 10001] -> [10000 10001] no-op
>   loadgen-1 p39: [10001 10003] -> [10003 10001] preferred leader
>   loadgen-1 p40: [10003 10004] -> [10003 10004] no-op
>   loadgen-1 p41: [10004 10005] -> [10004 10005] no-op
>   loadgen-1 p42: [10005 10000] -> [10005 10000] no-op
>   loadgen-1 p43: [10002 10001] -> [10002 10001] no-op
>   loadgen-1 p44: [10000 10002] -> [10000 10002] no-op
>   loadgen-1 p45: [10001 10005] -> [10005 10001] preferred leader
>   loadgen-1 p46: [10003 10002] -> [10003 10002] no-op
>   loadgen-1 p47: [10004 10003] -> [10004 10003] no-op
>   loadgen-1 p48: [10005 10004] -> [10005 10004] no-op
>   loadgen-1 p49: [10002 10000] -> [10002 10000] no-op
>   loadgen-1 p50: [10000 10004] -> [10000 10004] no-op
>   loadgen-1 p51: [10001 10003] -> [10003 10001] preferred leader
>   loadgen-1 p52: [10003 10005] -> [10003 10005] no-op
>   loadgen-1 p53: [10004 10002] -> [10004 10002] no-op
>   loadgen-1 p54: [10005 10000] -> [10005 10000] no-op
>   loadgen-1 p55: [10002 10001] -> [10002 10001] no-op
>   loadgen-1 p56: [10000 10001] -> [10000 10001] no-op
>   loadgen-1 p57: [10001 10003] -> [10003 10001] preferred leader
>   loadgen-1 p58: [10003 10002] -> [10003 10002] no-op
>   loadgen-1 p59: [10004 10005] -> [10004 10005] no-op
>   loadgen-1 p60: [10005 10004] -> [10005 10004] no-op
>   loadgen-1 p61: [10002 10000] -> [10002 10000] no-op
>   loadgen-1 p62: [10000 10001] -> [10000 10001] no-op
>   loadgen-1 p63: [10001 10002] -> [10002 10001] preferred leader
> 
> Broker distribution:
>   degree [min/max/avg]: 4/4/4.00 -> 4/4/4.00
>   -
>   Broker 10000 - leader: 13, follower: 8, total: 21
>   Broker 10001 - leader: 0, follower: 22, total: 22
>   Broker 10002 - leader: 14, follower: 8, total: 22
>   Broker 10003 - leader: 15, follower: 6, total: 21
>   Broker 10004 - leader: 10, follower: 11, total: 21
>   Broker 10005 - leader: 12, follower: 9, total: 21
> 
> WARN:
>   [none]
> 
> New partition maps:
>   loadgen-1.json
>

**Comment by zrosenberg** [comment]
Tried it again with an entire AZ fed in, and that also works fine. Only unexpected thing was that it requires you to run a preferred leader election afterwards if you don't have auto leader election on

**Review by jamiealquiza — APPROVED** [review]

**Pull request merged**